conf.go

```diff
@@ -32,6 +32,46 @@ type Sourcer interface {
 	Source(fld Field) (string, bool)
 }
 
+const versionKey = `SVN`
+const descKey = `Desc`
+
+// Version provides the abitily to add version and description
+type Version struct {
+	SVN  string // String Version
+	Desc string
+}
+
+// Details provides output to display the application version and description on the command line.
+func Details(namespace string, v interface{}) (string, error) {
```

Comment:
Why is this exported? The Usage and String calls should provide this information.

Reply:
Ok, I see, in case they only want version.  This should not be called details. We already have a Usage function.

VersionDisplay might be better and more accurate.

I hate that I allowed an error to be returned on Usage :(

Wondering if everything can be done with ErrHelpWanted and Usage?
Can we have usage check if we are looking for version, usage or both?

I don't want the user to have to change this
```
if err == conf.ErrHelpWanted {
	usage, err := conf.Usage("SALES", &cfg)
	if err != nil {
		return errors.Wrap(err, "generating config usage")
	}
	fmt.Println(usage)
	return nil
}
```

Reply:
I think the problem is the conf.Parse and conf.Usage are stateless so they don't know about each other.

One option that I can see is to change this
```go
// conf.go:74
flag, err := newSourceFlag(args)
if err != nil {
    return err
}
```
into something like this
```go
flag, err := newSourceFlag(args)
if err != nil {
    if err == ErrVersionWanted {
        VersionDisplay(namespace, cfgStruct)
        os.Exit(0)
    }
    return err
}
```
We can't return error because it will then throw the error up to stop the flow and we can't return error nil because then it will continue the application flow. But this is a little forced I think.

Reply:
I don't want to break the API if possible. The idea is

If Parse sees a `--help` or `--version` then return `ErrHelpWanted`
In Usage, check to see which option(s) were provided and give the user that information.

it doesn't matter that it is stateless.

Reply:
But then we need to sent the args to the Usage also right? Because the Usage function doesn't know about what args were sent, only knows the fields from the cfg.

```go
// Knows about the flag -v/--version
conf.Parse(os.Args, "CRUD", &cfg)
// Only knows about the cfg.Version value
conf.Usage("SALES", &cfg)
```

Unless I'm missing something on the flow between Parse and Usage.

Reply:
Ohhh Shit!  :(

Well technically, if the version information is not empty (after parse) we have what we need.

Reply:
Yes but the conf.Usage don't know if the user called -v or -h :/

One option is to create a package scope variable, not exposed that is set to true when the user calls with -v and then the conf.Usage will know that it's a -v not a -h. This way we should then push ErrHelpWanted and the logic will be inside the conf.Usage.

I know package scope variables are not a good practice but since it's not exposed and the zero value is false it can be an option. What do you thing?

Reply:
No package level variables.

So, let's add version information to Usage
Make sure String() does not show version information

Add a `DisplayVersion` function and the `ErrVersionWanted` variable back.

Reply:
Ok I did that, now the Parse can throw ErrVersionWanted and ErrHelpWanted and the user can catch both. So to get the version we have to do two steps, first add the conf.Version to the configuration struct and set the values, second add the error handler for ErrVersionWanted.

If the conf.Version is not added then -v option will not be available, if its added but the ErrVersionWanted is not handled it will exit like another error. This way "old code" will not be affected.